backend/src/federation/index.ts

```diff
@@ -36,18 +32,60 @@ export const startDHT = async (
     logger.info(`keyPairDHT: publicKey=${keyPair.publicKey.toString('hex')}`)
     logger.debug(`keyPairDHT: secretKey=${keyPair.secretKey.toString('hex')}`)
 
+    const apiList = {
+      apiVersions: Object.values(ApiVersionType).map(function (apiEnum) {
+        return { api: apiEnum, url: CONFIG.FEDERATION_COMMUNITY_URL }
+      }),
+    }
+    logger.debug(`ApiList: ${JSON.stringify(apiList)}`)
+
     const node = new DHT({ keyPair })
 
     const server = node.createServer()
 
     server.on('connection', function (socket: any) {
       // noiseSocket is E2E between you and the other peer
       // pipe it somewhere like any duplex stream
-      logger.info(`Remote public key: ${socket.remotePublicKey.toString('hex')}`)
+      logger.info(`server on... with Remote public key: ${socket.remotePublicKey.toString('hex')}`)
       // console.log("Local public key", noiseSocket.publicKey.toString("hex")); // same as keyPair.publicKey
 
-      socket.on('data', (data: Buffer) => logger.info(`data: ${data.toString('ascii')}`))
-
+      socket.on('data', async (data: Buffer) => {
+        try {
+          logger.info(`data: ${data.toString('ascii')}`)
+          const json = JSON.parse(data.toString('ascii'))
+          if (
+            json.apiVersions &&
+            Object.prototype.toString.call(json.apiVersions) === '[object Array]' &&
+            json.apiVersions.length > 0 &&
+            Object.prototype.toString.call(json.apiVersions[0].api) === '[object String]' &&
+            Object.prototype.toString.call(json.apiVersions[0].url) === '[object String]'
+          ) {
+            const communities = new Array<DbCommunity>()
+
+            for (let i = 0; i < json.apiVersions.length; i++) {
+              const apiVersion = json.apiVersions[i]
+              let community = await DbCommunity.findOne({
```

Comment:
What about this?
https://stackoverflow.com/questions/46745688/typeorm-upsert-create-if-not-exist

Reply:
As i understand the answers in your link the upsert will not properly work for all databases and uses the .onConflict wich is marked as deprecated.
There is an alternative suggestion using the QueryBuilder with .onUpdate. This will create one Statement and work on mariadb with a SELECT-Statement afterwards for reading the changed entity again:
INSERT INTO `communities`
(
	`id`, 
	`public_key`, 
	`api_version`, 
	`end_point`, 
	`last_announced_at`, 
	`created_at`, 
	`updated_at`
)
VALUES (DEFAULT, ?, ?, ?, ?, DEFAULT, DEFAULT) 
ON DUPLICATE KEY UPDATE 
	`end_point` = VALUES(`end_point`), 
	`last_announced_at` = VALUES(`last_announced_at`) 
-- PARAMETERS: 
[
	"6c554f16417caa1ecc2099a2c3692553a8c0fa4d6ffdf4a4c319513181d174f0",
	"v1_0",
	"http://localhost:4003/api",
	"2022-11-30T20:15:03.894Z"
]

**But this .onUpdate solution is marked as deprecated too!**

So which solution will satisfy you:
- the **deprecated** one with an INSERT ON DUPLICATE- and a following SELECT-Statement
- or the current regular valid one of TypeORM-Entity with a leading SELECT- and an INSERT- or UPDATE-Statement